### PR TITLE
Fix performance of `mesolve` with QArrays

### PR DESCRIPTION
**Pull request by gautierronan**
Fixes the ~20% performance loss with the new `vector_field` introduced with QArrays.

For now, only the dense format is supported. In a following commit, I plan to introduce a `StackedQArray` class to handle stacks of SparseDIA and/or Dense arrays, when relevant. 

<details>
<summary>Benchmark code</summary>

```python
import timeit
import dynamiqs as dq
import jax.numpy as jnp

dq.set_device('cpu')
dq.set_layout(dq.dense)

# parameters
kappa_2 = 1.0
eps_z = 0.1
nbar = 4.0
num_tsave = 100
N = 24

# time evolution
alpha = jnp.sqrt(nbar)
gate_time = jnp.pi / (4 * alpha * eps_z)
tsave = jnp.linspace(0.0, gate_time, num_tsave)

# operators
a = dq.destroy(N)
i = dq.eye(N)

# Hamiltonian
H = eps_z * (a + dq.dag(a))

# jump operators
jump_ops = [jnp.sqrt(kappa_2) * (a @ a - nbar * i)]

# initial state
psi0 = dq.unit(dq.coherent(N, alpha) + dq.coherent(N, -alpha))

# options
options = dq.Options(progress_meter=None)

# run
dq.mesolve(H, jump_ops, psi0, tsave, options=options).states.data.block_until_ready()
%timeit -n10 -r7 dq.mesolve(H, jump_ops, psi0, tsave, options=options).states.data.block_until_ready()
```

Before: 147 ms ± 14.1 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
After: 111 ms ± 632 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
</details>

### CODE REVIEW

**Review by pierreguilmin — APPROVED** [review]
Great investigation!
